### ISSUE TITLE
Store rendered HTML

### ISSUE THREAD

**Issue by sizzlemctwizzle** [OpenUserJs/OpenUserJS.org#638]
``` js
var mongoose = require('mongoose');
var Schema = mongoose.Schema;

var renderedContentSchema = new Schema({
  content: String,
  model: String,
  _contentId: Schema.Types.ObjectId
});

var RenderedContent  = mongoose.model('RenderedContent', renderedContentSchema);

exports.RenderedContent = RenderedContent;
```

This is where output from [renderMd](https://github.com/OpenUserJs/OpenUserJS.org/blob/master/libs/markdown.js) can be stored, so that we only render markdown when it changes.

**Comment by Martii** [comment]
This might be considered a bad idea if our sanitizer discovers a security hole and we have to redo the entire database. There is already precedence for this occasion.

**Comment by sizzlemctwizzle** [comment]
I'll build in a routine to re-render all stored content. I'll even put a
button for it on the admin dashboard.

**Comment by Martii** [comment]
Perhaps a flag check based off last updated compared to last rerendering date could work too to let the users do it on revisitation of those affected pages could work too... but we don't currently store that information in the db.

**Comment by sizzlemctwizzle** [comment]
Then we'd still have a security hole  as long as the content hasn't
expired. With my process we could fix everything immediately. Plus less
re-rendering.

**Comment by Martii** [comment]
Not quite...  on access it would check lastupdate versus the lastforced and any visitor would trigger the rendering and rerender it... thus the infected content never gets displayed to anyone. Your methodology is less scalable and will take a very long time when the db is much larger... USO had downtime because of this e.g. hours.

**Comment by Martii** [comment]
Related to #81 as well... one big db redo is going to be traffic intensive.

**Comment by sizzlemctwizzle** [comment]
I see what you mean and I agree. This let's us re-render when we need to
rather than all at once.

**Comment by joeytwiddle** [comment]
If this collection is just a cache, you could wipe the entire collection if a rendering issue is discovered. No need to store/compare dates.

I agree re-rendering and caching on demand sounds more scalable!

**Comment by sizzlemctwizzle** [comment]
We'll keep the lastRendered date on the new model to re-render on demand if
the rendering logic changes.

**Comment by sizzlemctwizzle** [comment]
@joeytwiddle might be right. I don't think it takes long destroy an entire collection. Once the collection is gone, all content will be forced re-render on demand. Of course our inevitable need to use sharding may make this slower. I really don't know.

**Comment by Martii** [comment]
@sizzlemctwizzle 
That's a null check to see if it's even been rendered e.g. date check... lastchecked vs lastforced is a general analogy which can easily be compared to before time and the last time. e.g. null vs exists. "Wiping" will still take longer especially when considering all comments/issues/discussions are rendered as well and not just script homepages... our DB isn't that large at the moment but give it a few more years and the performance could be potentially adversely affected.

**Comment by Martii** [comment]
Another miscellaneous note as well... DB compaction... eventually wiping could potentially fragment the database even worse than an overwrite... we haven't touch on that yet since the change at https://github.com/OpenUserJs/OpenUserJS.org/commit/d64f754af5818301648354e32596ed922d683e95

Since OUJS is now on SSD's it might be prudent to check the warranty status with the host provider and how often they perform backups of our VPS, replacements, etc. and if they do any compression anywhere.
